display_list/display_list_unittests.cc

```diff
@@ -1584,5 +1620,211 @@ TEST(DisplayList, SaveLayerBoundsSnapshotsImageFilter) {
+};
+
+TEST(DisplayList, SaveLayerOneSimpleOpSupportsOpacityOptimization) {
+  SaveLayerOptions expected =
+      SaveLayerOptions::kWithAttributes.with_can_distribute_opacity();
+  SaveLayerOptionsExpector expector(expected);
+
+  DisplayListBuilder builder;
+  builder.setColor(SkColorSetARGB(127, 255, 255, 255));
+  builder.saveLayer(nullptr, true);
+  builder.drawRect({10, 10, 20, 20});
+  builder.restore();
+
+  builder.Build()->Dispatch(expector);
+  EXPECT_EQ(expector.save_layer_count(), 1);
+}
+
+TEST(DisplayList, SaveLayerNoAttributesSupportsOpacityOptimization) {
+  SaveLayerOptions expected =
+      SaveLayerOptions::kNoAttributes.with_can_distribute_opacity();
+  SaveLayerOptionsExpector expector(expected);
+
+  DisplayListBuilder builder;
+  builder.saveLayer(nullptr, false);
+  builder.drawRect({10, 10, 20, 20});
+  builder.restore();
+
+  builder.Build()->Dispatch(expector);
+  EXPECT_EQ(expector.save_layer_count(), 1);
+}
+
+TEST(DisplayList, SaveLayerTwoOverlappingOpsPreventsOpacityOptimization) {
+  SaveLayerOptions expected = SaveLayerOptions::kWithAttributes;
+  SaveLayerOptionsExpector expector(expected);
+
+  DisplayListBuilder builder;
+  builder.setColor(SkColorSetARGB(127, 255, 255, 255));
+  builder.saveLayer(nullptr, true);
+  builder.drawRect({10, 10, 20, 20});
+  builder.drawRect({15, 15, 25, 25});
+  builder.restore();
+
+  builder.Build()->Dispatch(expector);
+  EXPECT_EQ(expector.save_layer_count(), 1);
+}
+
+TEST(DisplayList, NestedSaveLayersMightSupportOpacityOptimization) {
+  SaveLayerOptions expected1 =
+      SaveLayerOptions::kWithAttributes.with_can_distribute_opacity();
+  SaveLayerOptions expected2 = SaveLayerOptions::kWithAttributes;
+  SaveLayerOptions expected3 =
+      SaveLayerOptions::kWithAttributes.with_can_distribute_opacity();
+  SaveLayerOptionsExpector expector({expected1, expected2, expected3});
+
+  DisplayListBuilder builder;
+  builder.setColor(SkColorSetARGB(127, 255, 255, 255));
+  builder.saveLayer(nullptr, true);
+  builder.saveLayer(nullptr, true);
+  builder.drawRect({10, 10, 20, 20});
+  builder.saveLayer(nullptr, true);
+  builder.drawRect({15, 15, 25, 25});
+  builder.restore();
+  builder.restore();
+  builder.restore();
+
+  builder.Build()->Dispatch(expector);
+  EXPECT_EQ(expector.save_layer_count(), 3);
+}
+
+TEST(DisplayList, NestedSaveLayersCanBothSupportOpacityOptimization) {
+  SaveLayerOptions expected1 =
+      SaveLayerOptions::kWithAttributes.with_can_distribute_opacity();
+  SaveLayerOptions expected2 =
+      SaveLayerOptions::kNoAttributes.with_can_distribute_opacity();
+  SaveLayerOptionsExpector expector({expected1, expected2});
+
+  DisplayListBuilder builder;
+  builder.setColor(SkColorSetARGB(127, 255, 255, 255));
+  builder.saveLayer(nullptr, true);
+  builder.saveLayer(nullptr, false);
+  builder.drawRect({10, 10, 20, 20});
+  builder.restore();
+  builder.restore();
+
+  builder.Build()->Dispatch(expector);
+  EXPECT_EQ(expector.save_layer_count(), 2);
```

Comment:
Sorry, I don't understand what do u want to test. I fount we call once `saveLayer` the `expector.save_layer_count` is 1.
Through the case title `NestedSaveLayersCanBothSupportOpacityOptimization` I think the save_count will be less than 2, cause it is called `SaveLayer Opacity Optimization`.

Reply:
Both saveLayers can inherit and transmit the opacity information from a calling context to their children. That is what is being tested.

Reply:
Do we use the save_layer_count the check if saveLayer can inherit and transmit the opacity?

Reply:
I'm wondering if these questions have to do with save/Layer optimizations that might be coming?

The test is just making sure that the interactions of a saveLayer that is trying to figure out if all of its children are compatible works correctly in the presence of a single child that is itself a saveLayer which will also do its own compatibility processing. i.e. do the 2 layers correctly navigate and analyze their own compatibility independently?

The count itself here isn't a primary part of the test, but more used as a watchdog to make sure that the entire list of asynchronously delivered flags that were expected were completely encountered. Primarily the `expector` will verify that it encounters each flag property in its expectations, in the right order. But, since that class/object is called *by* another object that controls the start/stop of the iteration, it has no "and you should be done now" call in which it can double-check "thank you, I got all the flags so we are successful!" So if it doesn't get called once per saveLayer, then something was wrong other than getting a mismatch on the flags - so the testing of the count at the end is simply saying "and did we get a matching call for every set of flags that I provided in the constructor?" The count is a simple way to note that we did see all of the flags we were expecting, but that same condition could have been tracked or asserted in different ways (i.e. checking an iterator against the end of the list, or using a "are you done with all of the flags?, etc.).

In retrospect, it probably should have been implemented as:
`bool is_complete() { return save_layer_count_ == expected_.size(); }`
since the specific count itself is not at issue as much as the fact that it must match the length of the expectations vector. Also, rather than a count, it might have used an iterator and so `{ return cur_ == end_; }` might have been an alternate implementation of "completeness" that wouldn't depend on the internal workings of the expector class.